federation/pkg/federation-controller/replicaset/replicasetcontroller.go

```diff
@@ -0,0 +1,371 @@
+	}
+	<-stopCh
+	glog.Infof("Shutting down ReplicaSetController")
+	frsc.replicasetDeliverer.Stop()
+	frsc.clusterDeliverer.Stop()
+	frsc.replicasetWorkQueue.ShutDown()
+}
+
+func (frsc *ReplicaSetController) isSynced() bool {
+	if !frsc.fedInformer.ClustersSynced() {
+		glog.V(2).Infof("Cluster list not synced")
+		return false
+	}
+	clusters, err := frsc.fedInformer.GetReadyClusters()
+	if err != nil {
+		glog.Errorf("Failed to get ready clusters: %v", err)
+		return false
+	}
+	if !frsc.fedInformer.GetTargetStore().ClustersSynced(clusters) {
+		return false
+	}
+	if !frsc.replicaSetController.HasSynced() {
+		glog.V(2).Infof("federation replicaset list not synced")
+		return false
+	}
+	return true
+}
+
+func (frsc *ReplicaSetController) deliverLocalReplicaSet(obj interface{}, duration time.Duration) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		glog.Errorf("Couldn't get key for object %v: %v", obj, err)
+		return
+	}
+	_, exists, err := frsc.replicaSetStore.GetByKey(key)
+	if err != nil {
+		glog.Errorf("Couldn't get federation replicaset %v: %v", key, err)
+		return
+	}
+	if exists { // ignore replicasets exists only in local k8s
+		frsc.deliverReplicaSetByKey(key, duration, false)
+	}
+}
+
+func (frsc *ReplicaSetController) deliverFedReplicaSetObj(obj interface{}, delay time.Duration) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		glog.Errorf("Couldn't get key for object %+v: %v", obj, err)
+		return
+	}
+	frsc.deliverReplicaSetByKey(key, delay, false)
+}
+
+func (frsc *ReplicaSetController) deliverReplicaSetByKey(key string, delay time.Duration, failed bool) {
+	if failed {
+		frsc.replicaSetBackoff.Next(key, time.Now())
+		delay = delay + frsc.replicaSetBackoff.Get(key)
+	} else {
+		frsc.replicaSetBackoff.Reset(key)
+	}
+	frsc.replicasetDeliverer.DeliverAfter(key, nil, delay)
+}
+
+func (frsc *ReplicaSetController) worker() {
+	for {
+		item, quit := frsc.replicasetWorkQueue.Get()
+		if quit {
+			return
+		}
+		key := item.(string)
+		err := frsc.reconcileReplicaSet(key)
+		frsc.replicasetWorkQueue.Done(item)
+		if err != nil {
+			glog.Errorf("Error syncing cluster controller: %v", err)
+			frsc.deliverReplicaSetByKey(key, 0, true)
+		}
+	}
+}
+
+func (frsc *ReplicaSetController) schedule(frs *extensionsv1.ReplicaSet, clusters []*fedv1.Cluster,
+	expected map[string]int64, actual map[string]int64) map[string]int64 {
+	// TODO: integrate real scheduler
+
+	plnr := frsc.planner
+	frsPref, _ := parseFederationReplicaSetReference(frs)
+	if frsPref != nil { // create a new planner if user specified a preference
+		plnr = planner.NewPlanner(frsPref)
+	}
+
+	replicas := int64(*frs.Spec.Replicas)
+	var clusterNames []string
+	for _, cluster := range clusters {
+		clusterNames = append(clusterNames, cluster.Name)
+	}
+	scheduleResult := plnr.Plan(replicas, clusterNames)
+	// make sure the return contains clusters need to zero the replicas
+	result := make(map[string]int64)
+	for clusterName := range expected {
```

Comment:
the clusters above is to extract the names as a string array

the expected is to handle the case that the planner gave a different result like the following:
A:2/B:3/C:4 => A:4/C5
I will have to format it into A:4/B:0/C:5 so that I can decrease replicas in cluster B to 0
An alternate way is to iterate over all the clusters regardless